File: go/specimen/tree.go
package specimen

import (
	"log"

	"github.com/ditrit/specimen/go/specimen/focustree"
)

// This file implement the focustree.Node interface for the NoduleRoot and for Nodules

// NoduleRoot implements focustree.Node

func (t NoduleRoot) GetChildren() (children []focustree.Node) {
	for _, c := range t {
		children = append(children, c)
	}
	return
}

func (NoduleRoot) IsLeaf() bool {
	return false
}

func (NoduleRoot) GetFlag() focustree.FlagType {
	return focustree.None
}

func (NoduleRoot) Warning(info string) {
	log.Printf("Warning: NoduleRoot: %s\n", info)
}

// Nodule implements focustree.Node

func (n Nodule) GetChildren() (children []focustree.Node) {
	// convert []Nodule to []focustree.Node
	for _, c := range n.Children {
		children = append(children, c)
	}
	return
}

func (n Nodule) IsLeaf() bool {
	return !n.HasContentKey
}

func (n Nodule) GetFlag() focustree.FlagType {
	return n.Flag
}

func (n Nodule) Warning(info string) {
	log.Printf("Warning(%s): %s\n", n.GetLocation(), info)
}

File: rust/src/tree.rs
use crate::nodule::Nodule;
use specimen__focustree as focustree;
use specimen__writable::Writable;
use std::io::Write;

// This file implements the focustree::Tree trait for the Nodule struct and the NoduleRoot type.

impl<'a> focustree::Tree<Nodule<'a>> for Nodule<'a> {
    fn is_leaf(&self) -> bool {
        self.is_leaf
    }

    fn get_flag(&self) -> focustree::Flag {
        self.flag
    }

    fn get_children(&self) -> Vec<&dyn focustree::Tree<Nodule<'a>>> {
        self.children
            .iter()
            .map(|n| n as &dyn focustree::Tree<Nodule>)
            .collect()
    }

    fn get_value(&self) -> Nodule<'a> {
        (*self).clone()
    }

    fn warning(&self, message: &str, stdout: &mut Writable) {
        let res = write!(stdout, "Warning({}): {}\n", self.get_location(), message);
        if let Err(e) = res {
            panic!("Error: {}", e);
        }
    }
}
